document-store/src/main/java/org/hypertrace/core/documentstore/mongo/collection/MongoCollectionOptionsApplier.java

```diff
@@ -0,0 +1,47 @@
+package org.hypertrace.core.documentstore.mongo.collection;
+
+import static org.hypertrace.core.documentstore.model.options.DataFreshness.SYSTEM_DEFAULT;
+import static org.hypertrace.core.documentstore.mongo.collection.MongoReadPreferenceConverter.convert;
+
+import com.mongodb.BasicDBObject;
+import com.mongodb.ReadPreference;
+import com.mongodb.client.MongoCollection;
+import java.util.HashMap;
+import java.util.Map;
+import lombok.Builder;
+import lombok.Value;
+import lombok.experimental.Accessors;
+import org.hypertrace.core.documentstore.model.config.ConnectionConfig;
+import org.hypertrace.core.documentstore.model.options.QueryOptions;
+
+public class MongoCollectionOptionsApplier {
+  private final Map<CacheKey, MongoCollection<BasicDBObject>> collectionCache;
+
+  public MongoCollectionOptionsApplier() {
+    this.collectionCache = new HashMap<>();
+  }
+
+  public MongoCollection<BasicDBObject> applyOptions(
+      final ConnectionConfig connectionConfig,
+      final QueryOptions queryOptions,
+      final MongoCollection<BasicDBObject> collection) {
+    final CacheKey cacheKey =
+        CacheKey.builder().readPreference(readPreference(connectionConfig, queryOptions)).build();
+    return collectionCache.computeIfAbsent(
+        cacheKey, key -> collection.withReadPreference(key.readPreference()));
```

Comment:
How is the CacheKey associated with the collection? Since ReadPreference can be either primary or secondary, does this mean that `this.collectionCache` contains a maximum of 2 values?

Reply:
> Since ReadPreference can be either primary or secondary, does this mean that this.collectionCache contains a maximum of 2 values?

Yes.